### ISSUE TITLE
Inventory check with scanner

### ISSUE THREAD

**Issue by daN4cat** [jekkos/opensourcepos#157]
There are on the market barcode scanners that store all the scanned barcodes for an inventory check.
Stock control is part of what you would need to do at the end of the fiscal year to report correctly the capital worth of your company.

The scenario would be as follow:
- The shop assistant scan all the items in the shop
- A file is extracted from the bar code scanner in .csv format (or converted to that format)
- The .csv file is uploaded to OSPOS
- OSPOS does an automatic inventory cross-check
- OSPOS reports any misalignment between missing items and wrong quantities

**Comment by jekkos** [comment]
good idea, I like this one. We had to do a manual stock count but took another approach, which was trying to correct the inventory levels manually upon scanning an item at hand.

a functionality that would allow one to add barcodes to a separate file and then show the differences would be a huge time saver for this job.

**Comment by daN4cat** [comment]
The approach you describe is what my wife's shop do as well and every girl in the shop hates :-)

Stock control with those portable scanners is far easier, walking around the shop and stock and assuming they don't miss or double count something. But if an issue pops up when you do the automatic check you go and recheck.

Most of the issues (outside the couple of shoplifting that happens every year) are selling one item in let say size S instead of size M because the barcode didn't scan and typed things manually (forgetting about the last part that is the size in the name :-) ), or they inserted the wrong quantity or they forgot to add an item. Latter one is not uncommon too.

**Comment by daN4cat** [comment]
Wild idea...

what if we had an Android application doing the barcode scan and communicating with OSPOS database with a restful API?

I used this app https://github.com/zxing/zxing on my Android phone to test the generated barcodes.

**Comment by jekkos** [comment]
yes should not be too hard to setup in this case. I've been looking into rest_contoller yesterday. 
however I wonder what the flow of actions would be in the android app. The barcode scanner is mostly one separate activity, so not sure if you'd have to close that activity it first to send out the last scanned barcode and reopen it to scan a new one.

**Comment by daN4cat** [comment]
The .csv file would be the simple solution, no time for Android fancy app. Maybe some other contributor could help if we find more willing to.

**Comment by daN4cat** [comment]
And if you play with https://github.com/zxing/zxing Android app, you can see that it's a continuous scanning, once it detects a barcode it beeps and shows the barcode.
In our case the beep will do and push the data to remote and move to the next automatically.

The App should open and close a check session on OSPOS and maybe register in OSPOS the check session? I don't know yet, I guess there could be different options here to figure out.

**Comment by jekkos** [comment]
The app should indeed provide a couple of options that allow the user to login to the system. Once this is done, login can happen automatically (as with most apps). I would properly logout and login upon starting/closing the app (for security reasons)

**Comment by jekkos** [comment]
Looks like CI3 has some nice REST extensions. Regarding a possible redo of the application, tables might be easier to do using REST + react.js bootstrap table component (providing pagination, bootstrap layout, all sorts of filtering options, by just configuring a couple of things)

http://torbjornzetterlund.com/how-to-work-with-restful-services-in-codeigniter-3-x/
https://github.com/AllenFang/react-bootstrap-table

**Comment by daN4cat** [comment]
Yes, I thought about having a RESET extension to do some data fetching from some external app, adding a secure token as authentication.

There are cases where you maybe just want to check from the mobile the current total amount just with an icon tap and small app. Or the list of what sold today. Going through the web interface is cumbersome.

Also you could fetch data from Google sheet and insert it automatically for some financial analysis.

**Comment by daN4cat** [comment]
For RESTful API development we could consider using https://apiary.io/

**Comment by chandrawg** [comment]
any update for this feature ?

**Comment by alvin101** [comment]
It is now labeled Epic, with means it is on the road map for the future.  you can always chip in with some dev resources, donations or coding....

**Comment by chandrawg** [comment]
how much i have to pay for this feature ?

**Comment by alvin101** [comment]
Anyone experienced in PHP programming can provide custom development for an open source project.  You can google for developers and ask for a bid. But maybe someone here will come forward.

**Comment by daN4cat** [comment]
It's more than that. This is a two part story.

**Comment by jekkos** [comment]
You can place bids now through bounty source.

**Comment by jekkos** [comment]
I'm writing on the wiki how we currently do this. It's a more or less manual process but it works. Willl also add some SQL queries there for reporting purposes.

**Comment by darkjedeye** [comment]
Hi There.... any update on this feature? I am busy studying flutter and dart, and once done... would like to build an app for this which would handle mobile scanning through the api... bt would need to know if you guys have implimented as yet... as i would love to use that in the mean time...

**Comment by jekkos** [comment]
Hi @darkjedeye. Curerntly there is no such app for this. Feel free to propose something as it would certainly be a valuable addition. It could be something standalone on a smartphone / tablet with a simple interface that communicates with the server backend.

**Comment by objecttothis** [comment]
@daN4cat @jekkos I have been working with several RESTful APIs in the in-house Website CMS that I am writing code on.  It wouldn't be too difficult for us to write an API endpoint to do what you're talking about.  I think the API infrastructure should be written with many API's in mind.  i.e., facility to generate an API client key and set/limit permissions for that key.  Then that gives freedom for us to not only write an API for inventory from a smart phone, but also Item CRUD operations on products via RESTful API. I know I would use/want to build and API endpoint to update and create items from my website CMS.
